apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.IntegrationTests/Trace/GrpcTraceHeaderFactoryTest.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright 2018 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax.ResourceNames;
+using Google.Cloud.ClientTesting;
+using Google.Cloud.Diagnostics.Common.Trace;
+using Google.Cloud.Spanner.Admin.Instance.V1;
+using Google.Cloud.Trace.V1;
+using Google.Protobuf.WellKnownTypes;
+using System;
+using Xunit;
+using TraceProto = Google.Cloud.Trace.V1.Trace;
+
+namespace Google.Cloud.Diagnostics.Common.IntegrationTests.Trace
+{
+    public class GrpcTraceHeaderFactoryTest : IDisposable
+    {
+        private static readonly TraceIdFactory _traceIdFactory = TraceIdFactory.Create();
+        private static readonly TraceEntryPolling _polling = new TraceEntryPolling();
+
+        private readonly string _testId;
+
+        private readonly IConsumer<TraceProto> _consumer;
+        private readonly IManagedTracer _tracer;
+
+        private readonly Timestamp _startTime;
+
+        public GrpcTraceHeaderFactoryTest()
+        {
+            _testId = IdGenerator.FromDateTime();
+
+            _consumer = new GrpcTraceConsumer(TraceServiceClient.Create());
+            _tracer = SimpleManagedTracer.Create(_consumer, TestEnvironment.GetTestProjectId(), _traceIdFactory.NextId(), null);
+
+            _startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+        }
+
+        [Fact]
+        public void TraceOutgoing()
+        {
+            var spanName = EntryData.GetMessage(nameof(TraceOutgoing), _testId);
```

Comment:
Eek, I hadn't realized we'd need a separate CallSettings for each method. This is icky.
I wonder whether the new gRPC interceptor support will help... let's leave that until we've got it working though.

Reply:
Well, CallSettings are not needed per method, but per client instance, what we need per method is to create the proper span, whose name needs to be (according to instructions)  package.service/method. I don't think we have that info anywhere in the client libs for each of the methods, so it will fall on clients to get that info correctly. Agree, super icky.

Reply:
Ah yes, I see. I think [this from gRPC](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Method.cs) would really help - it's available in the ClientInterceptorContext, so an interceptor may well be the way forward here.

Reply:
Oh, interesting, I'll take a look at the interceptor then.